File: android/lca/src/main/java/com/dreampany/lca/data/misc/ExchangeMapper.java
package com.dreampany.lca.data.misc;

import com.dreampany.framework.misc.SmartCache;
import com.dreampany.framework.misc.SmartMap;
import com.dreampany.framework.util.DataUtil;
import com.dreampany.framework.util.TimeUtil;
import com.dreampany.lca.api.cc.model.CcExchange;
import com.dreampany.lca.data.model.Exchange;
import com.dreampany.lca.misc.ExchangeAnnote;

import javax.inject.Inject;

/**
 * Created by Hawladar Roman on 5/31/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
public class ExchangeMapper {

    private final SmartMap<String, Exchange> map;
    private final SmartCache<String, Exchange> cache;

    @Inject
    ExchangeMapper(@ExchangeAnnote SmartMap<String, Exchange> map,
                   @ExchangeAnnote SmartCache<String, Exchange> cache) {
        this.map = map;
        this.cache = cache;
    }

    public Exchange toExchange(CcExchange in) {
        if (in == null) {
            return null;
        }
        String id = DataUtil.join(in.getExchange(), in.getFromSymbol(), in.getToSymbol());
        Exchange out = map.get(id);
        if (out == null) {
            out = new Exchange();
            map.put(id, out);
        }
        out.setId(id);
        out.setTime(TimeUtil.currentTime());
        out.setExchange(in.getExchange());
        out.setFromSymbol(in.getFromSymbol());
        out.setToSymbol(in.getToSymbol());
        out.setVolume24h(in.getVolume24h());
        out.setVolume24hTo(in.getVolume24hTo());
        return out;
    }
}

File: android/crypto/src/main/java/com/dreampany/crypto/data/source/mapper/ExchangeMapper.kt
package com.dreampany.crypto.data.source.mapper

import com.dreampany.crypto.api.crypto.model.CryptoExchange
import com.dreampany.crypto.data.model.Exchange
import com.dreampany.crypto.data.source.pref.AppPref
import com.dreampany.framework.data.source.mapper.StoreMapper
import com.dreampany.framework.data.source.repo.StoreRepo
import com.google.common.collect.Maps
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by roman on 21/3/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class ExchangeMapper
@Inject constructor(
    private val storeMapper: StoreMapper,
    private val storeRepo: StoreRepo,
    private val pref: AppPref
) {
    private val exchanges: MutableMap<String, Exchange>

    init {
        exchanges = Maps.newConcurrentMap()
    }

    @Synchronized
    fun add(exchange: Exchange) = exchanges.put(exchange.id, exchange)

    @Synchronized
    fun getItems(inputs: List<CryptoExchange>): List<Exchange> {
        val result = arrayListOf<Exchange>()
        inputs.forEach { input ->
            result.add(getItem(input))
        }
        return result
    }

    @Synchronized
    fun getItem(input: CryptoExchange): Exchange {
        Timber.v("Resolved Exchange: %s", input.market);
        val id: String = input.let { it.market + it.fromSymbol + it.toSymbol }
        var out: Exchange? = exchanges.get(id)
        if (out == null) {
            out = Exchange(id)
            exchanges.put(id, out)
        }
        out.market = input.market
        out.setFromSymbol(input.fromSymbol)
        out.setToSymbol(input.toSymbol)
        out.price = input.price
        out.setVolume24h(input.volume24h)
        out.setChange24h(input.change24h)
        out.setChangePct24h(input.changePct24h)
        return out
    }

}